### PR TITLE
Add LazyWorkerInterface

### PR DESCRIPTION
**Pull request by danlapid**
This is a new WorkerInterface class that will be used in the internal PR, see comments to understand what it does.

### CODE REVIEW

File: src/workerd/io/worker-interface.h [hunk 1]
```diff
@@ -151,6 +151,102 @@ class WorkerInterface: public kj::HttpService {
 // for the promise, then invoke the destination object.
 kj::Own<WorkerInterface> newPromisedWorkerInterface(kj::Promise<kj::Own<WorkerInterface>> promise);
 
+template <typename Func>
+class FuturePromisedWorkerInterface final: public WorkerInterface {
+public:
+  FuturePromisedWorkerInterface(Func func): func(kj::mv(func)) {}
+
+  void resolveFunc() {
+    promise = func()
```
Comment (kentonv) [inline src/workerd/io/worker-interface.h:160]: You probably only want to overwrite the promise if it isn't already set?

File: src/workerd/io/worker-interface.h [hunk 2]
```diff
@@ -151,6 +151,102 @@ class WorkerInterface: public kj::HttpService {
 // for the promise, then invoke the destination object.
 kj::Own<WorkerInterface> newPromisedWorkerInterface(kj::Promise<kj::Own<WorkerInterface>> promise);
 
+template <typename Func>
+class FuturePromisedWorkerInterface final: public WorkerInterface {
+public:
+  FuturePromisedWorkerInterface(Func func): func(kj::mv(func)) {}
+
+  void resolveFunc() {
```
Comment (kentonv) [inline src/workerd/io/worker-interface.h:159]: Maybe `ensureResolve()`?

Should this be private?

File: src/workerd/io/worker-interface.h [hunk 3]
```diff
@@ -151,6 +151,102 @@ class WorkerInterface: public kj::HttpService {
 // for the promise, then invoke the destination object.
 kj::Own<WorkerInterface> newPromisedWorkerInterface(kj::Promise<kj::Own<WorkerInterface>> promise);
 
+template <typename Func>
+class FuturePromisedWorkerInterface final: public WorkerInterface {
```
Comment (kentonv) [inline src/workerd/io/worker-interface.h:155]: Name suggestion: `LazyWorkerInterface`

**Comment by danlapid** [comment]
@npaun @AdityaAtulTewari @harrishancock ping

File: src/workerd/io/worker-interface.h [hunk 4]
```diff
@@ -151,6 +151,105 @@ class WorkerInterface: public kj::HttpService {
 // for the promise, then invoke the destination object.
 kj::Own<WorkerInterface> newPromisedWorkerInterface(kj::Promise<kj::Own<WorkerInterface>> promise);
 
+template <typename Func>
+class LazyWorkerInterface final: public WorkerInterface {
+public:
+  LazyWorkerInterface(Func func): func(kj::mv(func)) {}
```
Comment (jasnell) [inline src/workerd/io/worker-interface.h:157]: Non-blocking nit: might consider marking this explicitly `KJ_DISALLOW_COPY_AND_MOVE` but not critical with having the KJ_ASSERT_NONNULLs

**Review by jasnell — APPROVED** [review]

File: src/workerd/io/worker-interface.h (hunk 4 above)
Comment (danlapid) [reply src/workerd/io/worker-interface.h:157]: Yea copies would be bad but they are implicitly disabled by virtue of having an Own member.
I don't see why moves should be a problem.

Comment (jasnell) [reply src/workerd/io/worker-interface.h:157]: Largely just an explicit hygiene thing for me... kind of a safeguard against someone changing any of the methods later and forgetting a `KJ_ASSERT_NONNULL`. But I'm being a bit extra paranoid :-)

**Pull request merged**